### ISSUE TITLE
struts 2.5 and struts jquery 4.0.1 grid loading

### ISSUE THREAD

**Issue by afattahi54** [struts-community-plugins/struts2-jquery#45]
To load a grid after submitting a form successfully, we used below submit tag.
````
<sj:submit id="submitForm"  onSuccessTopics="gridLoader" ...
````
and
````
$.subscribe("gridLoader", function(event, data) {       
    var recviedRequest =  event.originalEvent.request.responseJSON;
````
However in struts 2.5 and struts jquery 4.0.1, I get the
````
TypeError: event.originalEvent.request is undefined
````
Well I try to invest the event object to find if a new object is defined, but I could not find any thing. As struts jquery plkugin uses jquery 2 I wonder if this is some thing with jquery 2.

**Comment by jogep** [comment]
Yes 4.0.x is based on jQuery2. As far as I know this API was not changed, but I will investigate.

**Comment by gromanc** [comment]
pingback: http://stackoverflow.com/questions/37681552/struts-2-5-and-struts-jquery-4-0-1-grid-loading

**Comment by afattahi54** [comment]
The `onCompleteTopics` and `onSuccessTopics` have same definition in java doc. We are updating our code to work with `onCompleteTopics`, are we on the correct track ?! By the way what will be the differences.

**Comment by jogep** [comment]
You mean main differences between struts2 jquery 3.7.1 and 4.0.1?

**Comment by afattahi54** [comment]
Dear @jogep let me ask this first ... 
What is the difference of `onSuccessTopics` and `onCompleteTopics`?

**Comment by jogep** [comment]
onCompleteTopics are published after the request is completed independent of the success of this request. onSuccessTopics are only published if the request was successfully.

**Comment by flofourcade** [comment]
Seems to be an incompatibility of jquery form plugin with jquery >=1.12.1. See issue #495 of jquery form plugin on github

**Comment by flofourcade** [comment]
Made some tests and I came confirm the issue  with jquery form plugin. There are 2 possible workarounds to make the onSuccess callback working as expected:

- Specify a `dataType` attribute in your `<sj:submit />` tag. 

- Replace this 
`if (!options.dataType && options.target) {
            var oldSuccess = options.success || function(){};
            callbacks.push(function(data) {
                var fn = options.replaceTarget ? 'replaceWith' : 'html';
                $(options.target)[fn](data).each(oldSuccess, arguments);           
            });
        }`
by this
`if (!options.dataType && options.target) {
            var oldSuccess = options.success || function(){};
            callbacks.push(function(data, status, xhr) {
                var fn = options.replaceTarget ? 'replaceWith' : 'html';
                $(options.target)[fn](data).each(function(){
                      oldSuccess.call(this, data, status, xhr);
                });
            });
        }`
in js/plugins/jquery.forms.js. You'll have to minify this file too.

The jquery form plugin has no commits since a long time, I don't know if we can hope a new release that fixes this issue ...
